### ISSUE TITLE
Tiled gallery block: issues with captions

### ISSUE THREAD

**Issue by simison** [Automattic/jetpack#11794]
Tiled gallery block uses captions very similar to core gallery block. They have several issues and bugs we should solve before shipping captions to our customers.

This is a list of all the issues, bugs and problems around captions so that it's easier to re-design how captions should work.

- Links in captions look awful (https://github.com/Automattic/wp-calypso/issues/29729)
    <img src="https://user-images.githubusercontent.com/87168/50507868-f2b01700-0a87-11e9-9adc-b9f691af7c8b.png" width="100" alt=""/>

- Longer captions cover "remove image" button in the editor (#29706)
    <img src="https://user-images.githubusercontent.com/87168/50507860-e1670a80-0a87-11e9-8a6f-bee5ed5d7df3.png" width="100" alt=""/>
    <img src="https://user-images.githubusercontent.com/87168/50507915-1ecb9800-0a88-11e9-8e64-737ac04b3324.png" width="130" alt=""/>

- Captions don't look great on small tiles. We need to rethink how we adjust caption text size when tiles scale down for small screens. Captions in a tile might look great on desktop but then the gallery looks awful on mobile.
    <img src="https://user-images.githubusercontent.com/87168/50508034-c34dda00-0a88-11e9-8221-cae5ee673218.png" width="400" alt=""/>

- Writing captions for circle layout doesn't work.
    We had to disable captions for circle layout because, well:
    <img src="https://user-images.githubusercontent.com/87168/50508124-22abea00-0a89-11e9-8719-ed15cc29da2d.png" width="200" alt=""/>
    <img src="https://user-images.githubusercontent.com/87168/46087202-7f5de700-c1b2-11e8-8a6a-c3c99a194335.png" width="300" alt=""/>

- Captions are automatically pulled in from image meta. It's easy enough to just not write captions, but cleaning out automatically added captions from a lot of images, especially on smaller screens is really annoying or impossible:
    <img src="https://user-images.githubusercontent.com/87168/50508207-94843380-0a89-11e9-8681-2dfb7ff5b923.png" width="300" alt=""/>

- Users might expect captions to appear only on mouse hover, just like with old tiled gallery. That isn't too amazing with touchscreens, though.
    <img src="https://user-images.githubusercontent.com/87168/46085419-ad412c80-c1ae-11e8-9799-689c1771ab67.gif" width="300" alt=""/>

- Detecting if caption is empty using `RichText.isEmpty()` is buggy. Calypso-Gutenberg seems to pass an old version of RichText format (array) and our block expects new format (string), thus making it seem like caption isn't empty, causing caption UI to behave weirdly and causing block invalidations on save. We need to `RichText.toHTMLString( caption )`

- Adding files from Media picker wipes out previously added captions (https://github.com/Automattic/wp-calypso/issues/29557, p7jreA-21M-p2)

--- 

cc @MichaelArestad @scruffian @mapk

**Comment by jenhooks** [comment]
As of today, I'm not seeing captions showing up at all.  See live here: https://jenwpjm.mystagingwebsite.com/testing-jetpack-tiled-gallery-captions/

Reported in 2009353-zen. Based on https://github.com/Automattic/wp-calypso/pull/30014 -- sounds like this is something that is already known? Can I confidently tell the user this is something that will be fixed?

**Comment by simison** [comment]
Nope, there are no plans to add captions to tiled gallery block any time
soon.

**Comment by RCowles** [comment]
@simison thanks for clarifying! We received a report that captions no longer appear underneath the full-size image in Carousel. I was able to repro here:
https://ryancowles.mystagingwebsite.com/2019/05/tiled-gallery-block/

Note in the above example, each image has a Title, Description, Caption, and Alt Text value. Only the Title and Description fields show in Carousel.

Is this related/expected behavior as well?

**Comment by simison** [comment]
I'm not quite sure how carousel picks those up — @jeherve would you be more familiar with Carousel?

**Comment by jeherve** [comment]
Yep, that's a known issue / byproduct of how the gallery blocks store their data:
https://github.com/Automattic/jetpack/issues/11594#issuecomment-474273664

**Comment by RCowles** [comment]
Thank you both for the additional context! Please feel free to close this issue as you see fit.

**Comment by formosattic** [comment]
User request in #2056566-zen.

**Comment by cecoates** [comment]
User requested: #9543186-hc

**Comment by samiff** [comment]
Reported in 2166921-zen

Noting that the [current documentation](https://jetpack.com/support/jetpack-blocks/tiled-gallery-block/) still mentions hover-over captions as a feature.

**Comment by simison** [comment]
> Noting that the current documentation still mentions hover-over captions as a feature.

Thanks for the heads up!

p1562740031135900-slack-jpop-quill

**Comment by dromero20** [comment]
Reported today in 13928682-hc

**Comment by KokkieH** [comment]
Another request for captions in the tiled gallery block in https://wordpress.org/support/topic/jetpack-image-gallery-carousel-updated/

**Comment by mriyazuddin** [comment]
Another user looking for captions in the Tiled Gallery Block: 2396123-zen

**Comment by galakhyati** [comment]
Another user looking for captions in Tiled Gallery block: #15807426-hc

**Comment by galakhyati** [comment]
@zyousafi and I replicated the issue on simple as well as on AT sites:
Test sites:

Premium: https://khyatipremium.wordpress.com/tiled-gallery-block/
Free: https://khyatifree.wordpress.com/tiled-gallery-page/
AT: https://mywptest.blog/tiled-gallery/

**Comment by KokkieH** [comment]
A request for hover-over captions in https://en.forums.wordpress.com/topic/mouse-hover-mosaic-image-gallery/

**Comment by supernovia** [comment]
Another user having trouble here.
https://en.forums.wordpress.com/topic/adding-captions-to-pictures-in-gallery/#post-3397985

What if we gave folks options for caption positioning?  Like 
Desktop: hover/hide/top/bottom/underneath
Mobile: hide/top/bottom/underneath

**Comment by darnelldibbles** [comment]
Just noting another request: https://wordpress.org/support/topic/jetpack-blocks/#post-12198301

**Comment by wpericam** [comment]
Unable to see image captions in a tiled gallery block (with tiled mosaic style applied). 2655712-zd

**Comment by wpericam** [comment]
Unable to see image captions in a tiled gallery block zd-2662402

**Comment by mnelson4** [comment]
Also encountered this today.

**Comment by damiannep** [comment]
Report of captions missing in tiled gallery block in 18210248-hc

**Comment by mnelson4** [comment]
As a user of this, these are the solutions I'd suggest:

* Use a light background instead of a dark one for captions (I thought that's what the old Tiled gallery did anyway). Then you can leave the hyperlink colors etc the same as usual.
* Limit caption size to 80% of the image size or something; if the caption is too long to fit, allow scrolling (I know nobody loves scrolling, but it's more usable than cutting off the text entirely)
* For circle layouts, just use the same light background color and otherwise treat the same. Don't let this hold up a popular feature.
* Continue pulling captions from image meta. For me that's a feature not a bug. Re-typing image captions (for those who want them) is far more annoying for than deleting them (for those who don't want them).

**Comment by kriskarkoski** [comment]
Another report of missing captions on Tiled Gallery in 8677284-hc. We should remove the mention from our public-facing docs if it's not working currently. I spent quite a bit of time testing expecting it to work before finding this issue.

**Comment by olesyabrk** [comment]
The guidelines for this project were based on the user feedback, comments on this issue, and UX best practices:

- Add option to remove caption from mobile 
- Add caption to the bottom of Circle images
- Darken default opacity for the bottom of images
- Option to add caption under the image
- Captions are not hover over
- No links in captions

**Visual workflow** 
1 step:
- Removed Default Style, we don’t need to duplicate that info.
- Added toggle to remove captions on mobile

![Frame-1](https://user-images.githubusercontent.com/51215741/82958982-93e81900-9f6b-11ea-8029-d2064700671b.png)

2 step:
- Darkened bottom of the image for better visibility.
- Removed the ability to have links due to limited use and poor visibility.
- Added text alignment.

![Frame-2](https://user-images.githubusercontent.com/51215741/82959190-fb9e6400-9f6b-11ea-9be1-74051758ac4c.png)

3 step:
![Frame-3](https://user-images.githubusercontent.com/51215741/82959277-31dbe380-9f6c-11ea-8989-b6ad304d9a7a.png)

3 step (alt, captions under the image):
![Frame-4](https://user-images.githubusercontent.com/51215741/82959320-4c15c180-9f6c-11ea-816e-2f0413dfbc8e.png)

3 step (alt, round images):
- for round images, the only way to display captions is under the image

![Frame](https://user-images.githubusercontent.com/51215741/82959553-d5c58f00-9f6c-11ea-9be6-ddb7be616c00.png)

3 step (alt, multiple quotes):
![alt with move quotes under images](https://user-images.githubusercontent.com/51215741/82959679-2806b000-9f6d-11ea-8a8a-61c806d35a5e.png)

Mobile:
![X - 1](https://user-images.githubusercontent.com/51215741/82959687-2d63fa80-9f6d-11ea-9fcb-90b5d6f76632.png)

**Comment by rickybanister** [comment]
> Removed Default Style, we don’t need to duplicate that info.

I think the "default style" dropdown is a global setting allowing customers to always use the same variation (like always default to rounded images).

**Comment by olesyabrk** [comment]
> I think the "default style" dropdown is a global setting allowing customers to always use the same variation (like always default to rounded images).

Thanks for pointing that out, @rickybanister . In that case perhaps we can keep the dropdown and remove the duplicate visual styles since the user can adjust and preview these options via the block bar instead. @creativecoder is there a global settings pattern that blocks use, any idea why we're offering our users three places to adjust their styles in the same block?

**Comment by susanjanec** [comment]
Just noting a request in 21619595-hc

**Comment by creativecoder** [comment]
@olesyabrk Thanks for these designs! @pento @aaronrobertshaw @glendaviesnz and I took a look at these last week, and everything seems in order. It looks like you've followed very similar conventions to the Core gallery block for this, which is great. I think this is a good opportunity for us to take a look at what kind of shared functionality/code is possible for the caption functionality in a way that could be shared across multiple blocks (and what's available in Core that could be reused).

> @creativecoder is there a global settings pattern that blocks use, any idea why we're offering our users three places to adjust their styles in the same block?

I think you're referring to

- The styles picker in the sidebar
- The "Default style" dropdown
- The style picker in the block toolbar that hovers above the block

These are all implemented by Core when [styles are registered for a block](https://developer.wordpress.org/block-editor/developers/block-api/block-registration/#styles-optional). The Default style should cause additional blocks of the same type to use that style by default when they are inserted.

**Comment by olesyabrk** [comment]
@creativecoder unifying as much is possible would be fantastic. Adding my P2 about the ideal / MVP experience [here](https://explorersp2.wordpress.com/2020/05/26/unifying-gallery-blocks-into-a-one-robust-experience/).

**Comment by rickybanister** [comment]
I wanted to ask about this recommendation:

> Removed the ability to have links due to limited use and poor visibility.

I understand that there may be accessibility/visibility issues, but I can imagine a number of use cases for having links within captions for semantic purposes (like giving a photographer credit). 

Since we already allow that, it feels worth looking for opportunities to improve it vs removing it.

**Comment by olesyabrk** [comment]
@rickybanister my suggestion was to roll out without the links and then iterate based on user feedback. I looked at a number of user sites, and haven't seen links in captions at all. Mostly it's just folks documenting their adventures. They're the photographers. So I'm basing this on the assumption that the majority of users don't have links in their captions. 

I can spin up a few iterations with links though since you think it's worth exploring further.

**Comment by damiannep** [comment]
Since individual images in a gallery can't have custom links, HEs recommend that users add links to captions as a workaround. Removing the option to add links to captions will remove this workaround to an existing product gap/technical limitation.

**Comment by msilbers** [comment]
@damiannep I think many HEs just suggest using a CoBlocks gallery which does allow linking out from each image. As far as I can tell, it's a pretty rare way to do that, and I've almost never seen a link in a caption (it's not that easy to click that way and doesn't seem especially accessible).

**Comment by damiannep** [comment]
You're right @msilbers. I was thinking of the old workaround before Coblocks galleries. Sorry for the confusion.

**Comment by olesyabrk** [comment]
thanks for chiming it @msilbers! It makes sense. I still think we should go ahead and release without the links, but if we do to include links, I'd go with the high contrast, no color like this. The text bar would include a link option. 
![image](https://user-images.githubusercontent.com/51215741/84449016-12e28000-ac01-11ea-8dc3-578253c5b502.png)

Thoughts, @rickybanister ?

**Comment by LambertusM** [comment]
Hi olesyabrk. I cant see the display caption option in the tiled gallery options. Did I miss something?

**Comment by jeherve** [comment]
@LambertusM Those changes have not been implemented yet. We'll update this issue, and close it, once the changes are added to the Jetpack plugin.

**Comment by rickybanister** [comment]
That works for me @olesyabrk!

**Comment by LambertusM** [comment]
@jeherve Thx. So, will need a bit more patience.

**Comment by pento** [comment]
I've just been reviewing this issue, I think it can potentially be achieved through deprecating `jetpack/tiled-gallery` in favour of block style variations registered against `core/gallery`.

There are four styles to be concerned about:

- _Circles_ and _Squares_: Both are quite similar to `core/gallery`, except that the width and height is the same. This could potentially be done with just CSS, though there'd probably be some hackery involved in calculating the height based on the width.
- _Columns_: This can be done entirely with CSS: set the `column-count` on `.blocks-gallery-grid` appropriately, remove flex from it, and change the `width` on `.blocks-gallery-item` to `100%`.
- _Mosaic_:  This is a trickier one, but I think it could be done using grid layout instead of flex. Playing around with it a little bit, something like this is showing promise:

```css
.blocks-gallery-grid {
	display: grid;
	// Make the grid 6 columns wide, each column equal width.
	grid-template-columns: repeat(6, 1fr);
	grid-gap: 16px;
	// Allow the layout engine to fill in gaps at the end.
	grid-auto-flow: dense;
}

// By default, images cover two columns, one row.
.blocks-gallery-item {
	width: 100%;
	height: 100%;
	grid-column: span 2;
}

.blocks-gallery-item figure {
	display: block;
}

.blocks-gallery-item img {
	// Images should expand to fit the whole block.
	width: 100%
	height: 100%;
	// Trim the edges of the image, if needed to fit.
	object-fit: cover;
}

// Every fourth image should be twice as big as normal.
.blocks-gallery-item:nth-of-type(4n) {
	grid-column: span 4;
	grid-row: span 2;
}

// Every eight image should be twice as big, and right aligned.
.blocks-gallery-item:nth-of-type(8n) {
	grid-column: 3 / span 4;
}
```

Tweaking or adding `nth-of-type` rules allows for more variety in the layout. I'd advise not adding too many rules that give big images, as there'll be not enough small images left over to fill the gaps at the end.

**Comment by glendaviesnz** [comment]
Had a look at some other options to reduce the code duplication between core and jetpack for this, like abstracting out some of the core gallery block into `@wordpress/components` so it could be shared by the jetpack blocks, but @pento idea seems much more promising so will try and get a PoC of that approach together.

**Comment by glendaviesnz** [comment]
Have added a very basic PoC at https://github.com/Automattic/jetpack/pull/16476 of adding a style to the core gallery ... it seems to work pretty effectively so worth pursuing this option further as this will give us all the caption functionality from core rather than having to duplicate it.

**Comment by pento** [comment]
WordPress/gutenberg#23929 is a similarly experimental PoC, but approaching from a slightly different direction: looking at generating mosaic rules based on the number of columns in the grid. Between that, and #16476, I think these are good indications that we're onto something.

**Comment by glendaviesnz** [comment]
> but approaching from a slightly different direction: looking at generating mosaic rules based on the number of columns in the grid. Between that

Ah, nice, that was my job for today to look at how we modify this to copy with multiple columns, I am going to close my PR in favour of your one for now.

**Comment by pento** [comment]
After a couple of days of experimenting with this approach, I have a few concerns, that I suspect will be difficult to address:

- Without having full control over the packing algorithm used to lay out the images within the gallery (as Jetpack galleries do), it's very difficult to make things line up exactly. I'm going to dig into the math a bit and see what I can come up with, but there's a high risk of this turning into a much more complex approach than I initially thought.
- Transitioning from the current Jetpack gallery to the new one will definitely require the end user to choose to upgrade: there are benefits to make the upgrade worth it, but due to different packing being used, the new gallery will inevitably look different to the current gallery, so we can't upgrade without the user deciding to make the change.
- As with anything, doing this in the Gutenberg plugin means we're 6+ months away from all Jetpack users being able to transition to the new gallery style.

While discussing the problem with @jasmussen, he noted that Core galleries will be changing their structure in the near future (WordPress/gutenberg#11436): they'll become container blocks, with standard image blocks used as the images. This approach is a potential option for Jetpack, too: use the standard image block in Jetpack galleries, which gives us the extra features we're after (in this case, captioning), while also reducing the complexity of Jetpack's code, and future maintenance burden.

I think it's worth doing an experimental attempt of this approach, as the potential benefits are a legit upside. If it turns out that this approach also has unforeseen drawbacks, then I think we'll have reached the point where we should fall back to implementing caption support within the current Jetpack gallery codebase, and leave bringing Jetpack and Core closer together for a future iteration.

**Comment by jasmussen** [comment]
I'm back from a little much appreciated AFK time, let me know if there's anything I can do to help here!

**Comment by otpidusprime** [comment]
User reported this feature to be missing in 3425228-zen

**Comment by fstath** [comment]
A forum user reported it:
https://wordpress.com/forums/topic/caption-in-tiled-block-group/

**Comment by supernovia** [comment]
Another request [here](https://wordpress.com/forums/topic/block-vs-classic/page/3/#post-3581158):

> why can’t tiled galleries have captions.

**Comment by sanjeev00733** [comment]
Another report in 25443273-hc

>  can you tell me why my captions don't appear when I click the photo on the website

**Comment by eduardozulian** [comment]
Another in 25477790-hc

**Comment by JoshBDunn** [comment]
Another request for this feature in [25682278-hc](https://mc.a8c.com/support-stats/happychat/transcript.php?id=25682278)

**Comment by allykmyta** [comment]
Another request for this feature in 26341980-hc

**Comment by chaitanyamsv** [comment]
26396768-hc

**Comment by Agepeuve** [comment]
Another in zen-3642369

**Comment by ahmadbaig1** [comment]
Tiled Gallery (Mosaic) doesn't show captions in slideshow mode.
User report- 3624123-zen

**Comment by TrovenTrip** [comment]
Hi all, I've come across this thread while trying to add captions to my JetPack tiled gallery blocks. Looks like there's been some good work on it so far, but nothing posted since July 2020. Just wondering if this is still being looked at? I'm happy to help with any beta testing on my site if/when a solution reaches that stage.
Thanks

**Comment by creativecoder** [comment]
Hi @TrovenTrip! I don't think anyone is actively working on this, specifically. There's [related work on the image gallery block](https://github.com/WordPress/gutenberg/pull/25940) in WordPress Core. The hope is that once that approach is proven, the same approach could be copied in the Tiled Gallery block. This would bring with it all the standard image block controls (including captions).

**Comment by noteeth** [comment]
Another request for captions in Tiled gallery
3782362-zen
27450396-hc

**Comment by Robertght** [comment]
Another request in 3785382-zen

**Comment by otpidusprime** [comment]
Another report in 27450396-hc
Follow up at 3782362-zen

**Comment by tmmbecker** [comment]
3788256-zen

**Comment by aleone89** [comment]
Another request for this has been made here: https://wordpress.com/forums/topic/find-pictures-captions-in-my-slider/

**Comment by kelasante** [comment]
Another request 28495746-hc

**Comment by lizkarkoski** [comment]
Another request from the forums today: https://wordpress.com/forums/topic/captions-13/
suggested they use the regular gallery block as a work around for now.

**Comment by retnonindya** [comment]
Request from a user on chat:
28274850-hc

I suggested them to use regular Gallery block.

**Comment by supernovia** [comment]
A request here for some control over gallery captions:
https://wordpress.com/forums/topic/galleries-out-of-control/

They'd like them to not show over the images, but show in the carousel view.

**Comment by katiebethbrown** [comment]
User expected captions to display on Tiled gallery block. Suggested they transform to core gallery instead.

29385594-hc

**Comment by msilbers** [comment]
requested in 30174287-hc
Recommended the core Gallery block instead

**Comment by retnonindya** [comment]
Another request:
30460629-hc

I suggested them to use core Gallery block

**Comment by Robertght** [comment]
New request in 30514763-hc

**Comment by jonnyruzek** [comment]
4133913-zen
Recommended the core Gallery block instead

**Comment by danjjohnson** [comment]
Requested in https://wordpress.org/support/topic/image-title-in-tiled-gallerys/

**Comment by tmmbecker** [comment]
Several Newspack publishers have asked about this. They love the tiled gallery but don't love that there are no captions :/

**Comment by daria2303** [comment]
30884014-HC requested here too

**Comment by rickmgithub** [comment]
User not seeing captions on the tiled gallery - 4194585-zd 
Seems to appear on desktop but not mobile.

**Comment by mikeicode** [comment]
Requested 31215860-hc

**Comment by mikeicode** [comment]
Another 31230051-hc

**Comment by kspilarski** [comment]
33575554-hc

**Comment by mxhassani** [comment]
Another one: 5372601-zd-woothemes

**Comment by filipanoscampos** [comment]
Another one:
5905506-zd-woothemes

**Comment by pehaa** [comment]
Another request in forum https://wordpress.org/support/topic/no-captions-in-tiled-gallery/

**Comment by erricgunawan** [comment]
+1 for my own personal request

**Comment by Robertght** [comment]
Another request made in https://github.com/Automattic/wp-calypso/issues/100662

**Comment by mxhassani** [comment]
Another request 9698933-zd-a8c

**Comment by mkstvns** [comment]
Really getting the sense this is a feature request we'll never see fulfilled, but, here's a vote from me too.

**Comment by jorpdesigns** [comment]
9956020-zd-a8c

**Comment by ariel-maidana** [comment]
Another request 10389588-zd-a8c
